tests/test_PDEModel.py

```diff
@@ -35,3 +37,320 @@ def test_model_input():
+        fenics_continuous_geo, length_scale=.1, num_terms=5)
+
+    # Create range geometry
+    range_geometry = cuqipy_fenics.geometry.FEniCSContinuous(
+        poisson1.solution_function_space)
+
+    # Create cuqi forward model (two models corresponding to two PDE objects)
+    cuqi_model1 = cuqi.model.PDEModel(
+        PDE1, domain_geometry=domain_geometry, range_geometry=range_geometry)
+
+    cuqi_model2 = cuqi.model.PDEModel(
+        PDE2, domain_geometry=domain_geometry, range_geometry=range_geometry)
+
+    # Create the prior
+    x = cuqi.distribution.Gaussian(mean=np.zeros(cuqi_model1.domain_dim),
+                                   cov=1, geometry=domain_geometry)
+
+    # Create exact solution and data
+    np.random.seed(0)
+    exact_solution = cuqi.samples.CUQIarray(
+        np.random.randn(domain_geometry.par_dim),
+        is_par=True, geometry=domain_geometry)
+
+    data1 = cuqi_model1(exact_solution)
+    data2 = cuqi_model2(exact_solution)
+
+    # Create likelihood 1
+    y1 = cuqi.distribution.Gaussian(mean=cuqi_model1(x),
+                                    cov=np.ones(cuqi_model1.range_dim)*.01**2)
+    y1 = y1(y1=data1)
+
+    # Create likelihood 2
+    y2 = cuqi.distribution.Gaussian(mean=cuqi_model2(x),
+                                    cov=np.ones(cuqi_model2.range_dim)*.01**2)
+    y2 = y2(y2=data2)
+
+    # Create posterior
+    cuqi_posterior = cuqi.distribution.JointDistribution(
+        y1, y2, x)._as_stacked()
+
+    # Solve the Bayesian problem
+    # Sample the posterior (Case 1: no reuse of assembled operators)
+    Ns = 20
+    np.random.seed(0)
+    sampler = cuqi.sampler.MetropolisHastings(cuqi_posterior)
+    t0 = time.time()
+    samples1 = sampler.sample_adapt(Ns, Nb=10)
+    t1 = time.time()
+    t_no_reuse = t1-t0
+    samples1.geometry = domain_geometry
+
+    if case == 1:
+        # Set PDE2 to be a shallow copy of PDE1 but with different rhs
+        PDE1.reuse_assembled = True
+        PDE2 = PDE1.with_updated_rhs(poisson2.rhs_form)
+        cuqi_model2.pde = PDE2
+
+        # Sample the posterior again (Case 2: reuse of assembled operators)
+        np.random.seed(0)
+        sampler = cuqi.sampler.MetropolisHastings(cuqi_posterior)
+        t0 = time.time()
+        samples2 = sampler.sample_adapt(Ns, Nb=10)
+        t1 = time.time()
+        t_reuse = t1-t0
+
+        # Check that the samples are the same
+        assert np.allclose(samples1.samples, samples2.samples)
+        
+        # Check that reusing factorization and with_updated_rhs is faster
+        assert t_reuse < 0.7*t_no_reuse
+
+    elif case == 2:
+        # Check that the samples matches the ones generated
+        # before updating the library code to add the reuse_assembled functionality
+        samples_orig = np.load(
+            copy_reference('data/samples_test_with_rhs_write_from_pytests'+\
+            '_0+untagged.69.ge349834.dirty.npz'))["samples1"]
```

Comment:
I think there is a bug in our version.py file when asking for version locally. Its not supposed to be written like that. I guess its OK for this file name, but it looks a bit wierd. Perhaps just keep the commit ID?

Reply:
I use the commit hash now.